### DISCUSSION TITLE
What steps to make JULIAdotNET work with .Net Framework ?

### DISCUSSION THREAD

**Discussion by BambOoxX** [HyperSphereStudio/JULIA.Net#13] [Q&A]
I am trying to adapt this into one of my existing .Net Framework projects, but it fails on the `UnmanagedCallersOnly`. 
I don't know how to modify the implementation in order to no use that, would you have an idea ?

**Comment by HyperSphereStudio** [comment]
Oops I forgot to push the latest version which added directives that included it if and only if the unmangaedcallersattribute existed.

**Reply by HyperSphereStudio** [reply]
I actually only use this project under .NET framework so that is the only thing that is Core specific. Only downside of using .net framework is you cant initialize .NET from julia


I will push the latest version later today

**Reply by BambOoxX** [reply]
Hmm, I just tested the updated code, now Visual Studio complains that about an `error CS8370` in `NativeSharp.cs` and `ObjectManager.cs`, did you try to compile against .Net Framework 4.7.2 or 4.8 ?

**Reply by HyperSphereStudio** [reply]
Could you post the entire error?


I havent yet released v3 yet. Because I now rely on the julia package I have to commit changes to .jl files to git in order to test them so the new updated code is not ready yet. 

I am working on not having to push to git to apply changes.


I am compiling with .net5.0. I will recompile my .net framework version and see if it works.

**Reply by BambOoxX** [reply]
Hmm, maybe there is an issue with the versioning then because Github says that the current version is already 3.0.0.

The errors I get are basically about features that are not available in C# 7.3 such as unmanaged constructed types as in `NativeArray`. 
In `ObjectManager` the error is manageable, it is due to the `new()` calls. Explicitely specifying the constructor solves this.

**Reply by HyperSphereStudio** [reply]
Ah I see. I personally prefer programming in c#10. 

Are you able to compile c#10? I can provide my .net framework .dlls mext release if you want I dont think c# version effects the .net framework version required to run it but if it does I can change the stuff to a lower version

**Reply by HyperSphereStudio** [reply]
I am currently trying to fix a glitch with boxing julia values. For whatever reason I cant figure out how to pass immutable types as a value ptr (Any type).

**Reply by BambOoxX** [reply]
I don'tmind using C# 10, in general,  but I'm trying to integrate this Julia interface in one of my. Net 4.7.2 projects which is not compatible with c#10 AFAIK.  And this project is tied with other dependencies that cannot work under. Net 5

**Reply by HyperSphereStudio** [reply]
Oh interesting I did not know that. I will lower the language to 7.6

**Reply by BambOoxX** [reply]
FYI If you follow this link https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-version-history#c-version-73,  you'll see that from c# 8.0 the language targets specifically net core.  I assume that implicitly that means that net framework is not supported with newer versions

**Reply by HyperSphereStudio** [reply]
I was able to lower the language to 7.3 but unfortunately 3.0.0 broke a lot of the internals of the code during testing. I will iron out the bugs but it will take some time (college starting back up :/ ).

**Reply by HyperSphereStudio** [reply]
One feature that will be really nice in 3.0.0 is automatic object ownership though! Previously you had to use sharppin / free to manually pin objects. The newest version will behind the scenes handle this for you

**Reply by BambOoxX** [reply]
Oooooh that's a nice addition indeed,  well done!

**Comment by damian-666** [comment]
btw , but generally , I have have had to move everything to net6.. even if
WPF and only windows only and DirectX interop on some of it.   I build a
game with CCScipt, AvalonEdit, level editor WPF, C# and MONOgame wiht
dynamics code in it.  embedded in the wpf..

struggling for weeks but getting close.   because of legacy net4.8 modules, n
the level editor.

Microsfolt Code Analysis is a preparser that might support other
languages.   CSScript is c# only.   .Net Silk attempts to generate wrappers
for all graphics lib.   all are getting HEAVY daily flux from 100s of
people, some sponsored, some on MSFT payroll, and some on JetBrains, its
amazing they work together under Net foundation with only the slightest
conflict of interest over tooling.

But MSFT is selling cloud services mostly now ...

as for moving to net6 , new core.  its just impossible not to..  the
industry is moving that way and you cant link.48 to net6 libs.     i have
legacy code but windows-upgrade helped a little.. i had to make all new
projects and use the new simpler SDK projects.. i strongly suggest you
ditch 4.8 completely,  its dead.. and netstandard 2.0 isnt realy working..
i had to upgrade  from netstandard 2 to net6 cores and my physics libraries
speed up 20%  .  im using Monogame, scripting , plugins, depubbign and
interested in scientific visualization... Net 6 is completely open source ,
there is an amazing Silk. Net project wth tons of MFST stafff oon it.   I
guess there are tooling squabbles but JetBrains is sponsoring as well..
Julia is a very interesting language for science but I use C# for its
incredible performance adn deployability..  I think julia if strongly typed
has a strong future in actual production code but,   for game engines adn
perfromance scientfic simulation, for targeting many platforms wth one
codebase, its looks like .net foundation is better broad support.. MSFT
isnt caring about locking anyone to any platform they make their revenue on
cloud services now.       merely updating form Monogame 3.8 to 3.81 has set
me back 3 weeks.. i just want clean simple project files that target
anything..  Some of the science code i i see now is Julia, like Gabriel
Weymouth's CFD code lilypad and frankly i thnk that its the best syntax.  But
i need to deploy production code that runs physics at 4000 fps and draws at
least 60hz.


simply rebuilding my physics engine to net6 gave me 40% frame rate boost..
its worth it.  I dont know how julia works.. how it deploys but
compatibility with .net foundation somehow would be great contribution to
science.   Right now most simulationcode is openGL, cuda or stuff that only
one person can compile.  In dotnet tooling you clone build and it works. (
or should)  and on almost any device with multiple tools, Resharpers, MS
Code, Dev Studio, running on ARM, deploying on phones, MS Code etc...

sorry dont have time make this clear and concise m swamped with unexpectged
work but it has to go forward even now when its in flux...  it will
stabilize after this major transition i hope

**Reply by BambOoxX** [reply]
I'm not sure if I understood your point fully, but just to rephrase the .NET Framework issue in my case : we build our soft partially on top of a commercial soft that, at the moment, is locked on 4.7.2. I would like for the company responsible for that soft to upgrade, but it's not planned at the moment. Otherwise, I'd be really happy to upgrade myself, it would surely save us from a lot of trouble.

**Reply by damian-666** [reply]
you might try MSFTs upgrade-assistant...  tool,  if helped me in my case by adding some windows compability stuf to legacy modules that need net48 to work or are no longer maintained..  now m using net6 on everything i have source to or open source thats been maintaned,  but for those dlls,  and it brings in the old net 4.8 stuff somehow, its working on windows at least.  probably worth a try...

**Reply by BambOoxX** [reply]
Thanks I'll have a look at that

**Reply by BambOoxX** [reply]
@damian-666, just tested the migration tool on another project. It's not a one-click business, but it's fair to say it's working just fine. Thanks for the tip.

**Comment by damian-666** [comment]
yea , glad i might have helped someone but dont get too optimistic unless
your actual project u need to ship works and deploys .        its depends
on yhour roadmap i guess..     it was like press 1, over and over and ten
min later it added tons of stuff and worked.  I had some good luck, at
first ..BUT...

  but in case its relevant to your similar struggle   .i strripped
everything the assistnat added..    because now  Im stuck because someone
dropped Netstandard 2 support in one of my components i   updated,
monogame, to mg 3.8..1 i probably should not have adopted it so soon..
 the upside is that it cana use ssystem numerics persistent Vector2  .using
SIMD.. which i  really want. and Netstandard 2 doesnt have it....       so
i moving ahead despite the cost.

so then i had to  i bite the bullet rebuild all my projects from stratch
templates,     moving ato net6 libs, net 6 core..... the files are soooo
simple now..i stripped the stuff upgrade assisten added and dont see
regression YET..    .  its all ANYCPU...  also , for now..     then
referenced the old l closed sourch icensed components  docking wpf dlls on
net48 and  most of it works. and  even w/o compatibility sutff.u .but ,
because of nuances other things are broken,   Im so stuck i am looking at
Silk.net , other UI frameworks, as well for fresh ideas. but this SIlk  is
very new.but very clever methodology they are using to do unification of
all the zillion device for scientific visualization.        anwyas glad if
that new MSFT  script actually helps in your case..  they update that tool
often..      The new project . net project format and tooling is much
simpler,  but well, completely different... msbuild has alot of new things
in config and json files one mght need ot read about..  It would be great
if julia and silk cna work together.. because i do simulations and physics
games with fluids and rigid bodies .

c# math doesnt look like math in the papers and its rare to find research
code that builds and links.       . and would love to have some solvers
that look like math from the research papers AND   shippable / scritpable
high performance code like net 6 is just blazing fast... and im running on
about 4 differnet targets now.. android and pc from shared code, and with
shaders...

 Also if weird things happen when u revert changes and using vs 2022 ,
CLEANyour cache files.. those would be  are in USER and in TEmp ..folers on
ur machine.   fi using. Vs 2022 , search by date and clean anything
thats went things went south after you revert something...     long read
but i spend 4 weeks on this.. so hope something saves u time.

**Reply by BambOoxX** [reply]
Well, thanks for all this info. I have to confess my use case requires much less interaction between Julia and dotnet and also much less 3d vizualization. Nevertheless I fully understand you troubles, good luck with all this!
